languages/ql/generic/QL_to_generic.ml

```diff
@@ -0,0 +1,455 @@
+module H = AST_generic_helpers
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+(* AST_ql to AST_generic.
+ *
+ * See AST_generic.ml for more information.
+ *
+ *)
+
+(*****************************************************************************)
+(* Helpers *)
+(*****************************************************************************)
+
+let id x = x
+let option = Option.map
+let list = List_.map
+let string = id
+let bool = id
+let unsafe_fake s = Tok.unsafe_fake_tok s
+let fb = Tok.unsafe_fake_bracket
+
+(*****************************************************************************)
+(* Entry point *)
+(*****************************************************************************)
+
+let info x = x
+
+let wrap _of_a (v1, v2) =
+  let v1 = _of_a v1 and v2 = info v2 in
+  (v1, v2)
+
+let bracket of_a (t1, x, t2) = (info t1, of_a x, info t2)
+let ident x = x
+
+let rec name = function
+  | Id id -> G.Id (ident id, G.empty_id_info ())
+  | IdQualified x -> G.IdQualified (qualified_info x)
+
+and qualified_info { name_last = v1, v2; name_middle } =
+  let name_last = (ident v1, option type_arguments v2) in
+  let name_middle =
+    Some
+      (G.QDots
+         (list (fun (x, y) -> (ident x, option type_arguments y)) name_middle))
+  in
+  G.{ name_last; name_middle; name_top = None; name_info = G.empty_id_info () }
+
+and type_argument (v1, pi_opt) =
+  let v1 = TyN (name v1) |> G.t in
+  match pi_opt with
+  | None -> G.TA v1
+  | Some pi ->
+      G.OtherTypeArg
+        ( ("TypeArgWithArity", unsafe_fake ""),
+          [ G.T v1; G.E (G.L (G.Int pi) |> G.e) ] )
+
+and type_arguments x = Tok.unsafe_fake_bracket (list type_argument x)
+and type_ x = G.TyN (name x) |> G.t
+and name_of_type_ x = name x
+
+and module_definition (_tok, id, _typarams, rhs) =
+  let ent = G.basic_entity (ident id) in
+  let def =
+    match rhs with
+    | ModuleBody stmts -> G.ModuleStruct (None, list stmt stmts)
+    | ModuleAlias qual_info ->
+        G.ModuleAlias
+          (IdQualified (qualified_info qual_info) |> H.dotted_ident_of_name)
+  in
+  DefStmt (ent, ModuleDef { mbody = def }) |> G.s
+
+and param_of_vardecl = function
+  | VardeclInit (ty, id) ->
+      let ty = type_ ty in
+      let id = ident id in
+      G.Param (G.param_of_id ~ptype:(Some ty) id)
+  | VardeclEllipsis tok -> G.ParamEllipsis tok
+
+and params_any_of_vardecls vardecls =
+  G.Params (List_.map param_of_vardecl vardecls)
+
+and predicate_definition (v1, v2, v3, v4) =
+  let v1 = option type_ v1 in
+  let ent = G.basic_entity (ident v2) in
+  let fparams = List_.map param_of_vardecl v3 |> fb in
+  let body =
+    match v4 with
+    | PredicateExpr None -> G.FBNothing
+    | PredicateExpr (Some e) -> G.FBExpr (expr e)
+    | PredicateAlias (qual_info, pi) ->
+        G.FBExpr
+          (G.OtherExpr
+             ( ("PredicateAlias", unsafe_fake ""),
```

Comment:
Similarly for these I think you could just use the token from v2. It would result in a slightly off location, but unsafe_fake can result in an error.

(I guess slightly off locations can also result in crashes if we end up calling a weird substring so it's still not ideal but less likely)